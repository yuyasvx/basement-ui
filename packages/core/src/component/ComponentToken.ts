export const ComponentToken = {
  MODAL_FRAME: 'bmui-modalframe',
  modalFrame: {
    BACKDROP: 'bmui-modalframe--backdrop',
    CONTAINER: 'bmui-modalframe--container',
    CONTENT: 'bmui-modalframe--content',
  },
  OVERLAY: 'bmui-overlay',
  CARD: 'bmui-card',
  ALIGNMENT: 'bmui-alignment',
} as const;
