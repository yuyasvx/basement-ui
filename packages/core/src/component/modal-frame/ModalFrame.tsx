import clsx from 'clsx';
import React, { forwardRef, type PropsWithChildren, type ReactNode, useCallback, useMemo, useRef } from 'react';
import { createPortal } from 'react-dom';
import { type ClassNameCustomizable } from '../../hook/custom-class/CustomClassHook';
import { useOverlayInitializer } from '../../hook/overlay/OverlayInitializerHook';
import { type AlignmentProps, useAlignment } from '../alignment/Alignment';
import { ComponentToken } from '../ComponentToken';

export type ModalFrameProps = PropsWithChildren<{
  backdropLock?: boolean;
  backdrop?: ReactNode;
  onBackdropClick?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
  enableAlignment?: boolean;
}> &
  ClassNameCustomizable &
  AlignmentProps;

function preventPropagation(evt: React.MouseEvent<HTMLElement, MouseEvent>) {
  evt.stopPropagation();
}

export const ModalFrame = forwardRef<HTMLDivElement, ModalFrameProps>((props, ref) => {
  const modalFrameClass = useMemo(() => clsx(ComponentToken.MODAL_FRAME, props.className), [props.className]);

  const { alignmentClassName } = useAlignment(props);
  const modalFrameContainerClass = useMemo(
    () =>
      clsx(
        ComponentToken.modalFrame.CONTAINER,
        { ['-full']: props.backdropLock && props.enableAlignment },
        { [alignmentClassName]: props.backdropLock && props.enableAlignment },
      ),
    [alignmentClassName, props.backdropLock, props.enableAlignment],
  );

  const { overlayElementId } = useOverlayInitializer();
  const target = document.getElementById(overlayElementId);
  const { children, backdropLock, backdrop } = props;
  const internalRef = useRef(null);
  const r = ref != null ? ref : internalRef;
  const cb = useCallback(
    (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      event.stopPropagation();
      props.onBackdropClick?.(event);
    },
    [props],
  );
  ModalFrame.displayName = 'ModalFrame';

  if (target == null) {
    return <></>;
  }

  return createPortal(
    <div className={modalFrameClass} ref={r}>
      {backdropLock && <div className={ComponentToken.modalFrame.BACKDROP}>{backdrop}</div>}
      <div className={modalFrameContainerClass} onClick={cb}>
        <div className={ComponentToken.modalFrame.CONTENT} onClick={preventPropagation}>
          {children}
        </div>
      </div>
    </div>,
    target,
  );
});
