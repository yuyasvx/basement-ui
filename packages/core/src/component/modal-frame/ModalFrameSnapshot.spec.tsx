import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { ComponentToken } from '../ComponentToken';
import { ModalFrame } from './ModalFrame';

describe('Modal Frame Snapshot', () => {
  test('Backdrop Lockが有効な時だけBackdropのカスタムができる', () => {
    render(
      <>
        <ModalFrame backdropLock backdrop={<div>backdrop</div>}>
          <div>modal content</div>
        </ModalFrame>
        <ModalFrame backdrop={<div>backdrop</div>}>
          <div>modal content</div>
        </ModalFrame>
      </>,
    );
    const portalTarget = document.getElementById(ComponentToken.OVERLAY);
    expect(portalTarget).toMatchSnapshot();
  });

  test('Backdrop Lockが有効な時だけ中身の位置調整が有効になる', () => {
    render(
      <>
        <ModalFrame backdropLock enableAlignment>
          <div>modal content</div>
        </ModalFrame>
        <ModalFrame enableAlignment>
          <div>modal content</div>
        </ModalFrame>
        <ModalFrame enableAlignment horizontalAlign="start">
          <div>modal content</div>
        </ModalFrame>
        <ModalFrame backdropLock enableAlignment horizontalAlign="start">
          <div>modal content</div>
        </ModalFrame>
      </>,
    );
    const portalTarget = document.getElementById(ComponentToken.OVERLAY);
    expect(portalTarget).toMatchSnapshot();
  });
});
