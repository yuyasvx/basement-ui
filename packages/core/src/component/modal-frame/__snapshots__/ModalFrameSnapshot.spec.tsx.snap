// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Frame Snapshot > Backdrop Lockが有効な時だけBackdropのカスタムができる 1`] = `
<div
  id="bmui-overlay"
>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--backdrop"
    >
      <div>
        backdrop
      </div>
    </div>
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal Frame Snapshot > Backdrop Lockが有効な時だけ中身の位置調整が有効になる 1`] = `
<div
  id="bmui-overlay"
>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--backdrop"
    >
      <div>
        backdrop
      </div>
    </div>
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--backdrop"
    />
    <div
      class="bmui-modalframe--container -full bmui-alignment"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--container"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
  <div
    class="bmui-modalframe"
  >
    <div
      class="bmui-modalframe--backdrop"
    />
    <div
      class="bmui-modalframe--container -full bmui-alignment -h-start"
    >
      <div
        class="bmui-modalframe--content"
      >
        <div>
          modal content
        </div>
      </div>
    </div>
  </div>
</div>
`;
